# Necessary Libraries

In [1]:
import numpy as np
from sympy import Matrix
import matplotlib.pyplot as plt
import seaborn as sns

**1. Create Your Personal Matrix**

In [2]:
student_id = "1056"
d1 = int(student_id[-1])   # last digit  -> 6
d2 = int(student_id[-2])   # second last digit -> 5

A = np.array([[d1 + 2, d2 + 1],
              [2*d1,   d2 + 2]])
print("Matrix A:\n", A)

Matrix A:
 [[ 8  6]
 [12  7]]


**2. Analyze Matrix A**

In [3]:
shape_A = A.shape
det_A = np.linalg.det(A)
rank_A = np.linalg.matrix_rank(A)
eigvals_A = np.linalg.eigvals(A)

print("Shape of A:", shape_A)
print("Determinant of A:", det_A)
print("Rank of A:", rank_A)
print("Eigenvalues of A:", eigvals_A)

if det_A != 0:
    inv_A = np.linalg.inv(A)
    print("Inverse of A:\n", inv_A)
else:
    print("Inverse of A: does not exist (determinant is 0)")

Shape of A: (2, 2)
Determinant of A: -16.000000000000007
Rank of A: 2
Eigenvalues of A: [16. -1.]
Inverse of A:
 [[-0.4375  0.375 ]
 [ 0.75   -0.5   ]]


**3. Value Change Experiment**

In [4]:
B = A.copy()
B[0, 0] = B[0, 0] + 1        #Element changed: A[0,0] (top-left, d1+2 = 8) → add 1 → 9
print("Matrix B:\n", B)

Matrix B:
 [[ 9  6]
 [12  7]]


**4. Re-analyze Matrix B**

In [7]:
shape_B = B.shape
det_B = np.linalg.det(B)
rank_B = np.linalg.matrix_rank(B)
eigvals_B = np.linalg.eigvals(B)

print("Shape of B:", shape_B)
print("Determinant of B:", det_B)
print("Rank of B:", rank_B)
print("Eigenvalues of B:", eigvals_B)

if det_B != 0:
    inv_B = np.linalg.inv(B)
    print("Inverse of B:\n", inv_B)
else:
    inv_B = None
    print("Inverse of B: does not exist (determinant is 0)")

Shape of B: (2, 2)
Determinant of B: -9.000000000000002
Rank of B: 2
Eigenvalues of B: [16.54400375 -0.54400375]
Inverse of B:
 [[-0.77777778  0.66666667]
 [ 1.33333333 -1.        ]]


# **5. Comparison & Explanation**

1. How did the determinant change and why?

It went from −16 to −9 — a change of +7. For a 2×2 matrix, det = ad − bc. Only a (top-left) changed, from 8 to 9, and d = 7, so the determinant shifted by exactly (+1)×7 = +7. This is a direct consequence of the determinant being linear in each row/column when you hold the others fixed — increasing one entry changes the determinant by (that entry's cofactor × the amount added).

2. Did the rank change? Explain.

No — rank stayed at 2 for both. Since neither determinant is zero, both matrices are full-rank (non-singular); a small perturbation to one element wasn't enough to make the rows/columns linearly dependent. Rank would only drop if the change happened to make one row a scalar multiple of the other.

3. How did the eigenvalues respond to the value change?

The eigenvalues also changed. For matrix A, they were [16., -1.], and for matrix B, they became [16.54400375, -0.54400375]. A small change in a matrix entry can lead to changes in its eigenvalues, as they are roots of the characteristic polynomial det(A - λI) = 0. The exact relationship is complex, but generally, perturbing a matrix will perturb its eigenvalues.

4. Is B easier or harder to invert than A? Why?

Neither matrix is significantly 'easier' or 'harder' to invert in terms of computational complexity, as both are 2x2 matrices and their inverses can be calculated directly. However, in terms of numerical stability, a matrix with a determinant closer to zero is generally considered 'harder' to invert because small changes in its entries can lead to large changes in its inverse (it is closer to being singular). Matrix A has a determinant of approximately -16, while matrix B has a determinant of approximately -9. Since the absolute value of the determinant of B (9) is smaller than that of A (16), matrix B is technically 'closer' to being singular than A. This suggests that B might be slightly more sensitive to numerical errors during inversion, making it arguably 'harder' or less stable to invert from a numerical perspective, though for these small, well-conditioned matrices, the difference is negligible.